src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs

```diff
@@ -230,14 +230,21 @@ private async Task UpdateSourceSymbolTreeInfoAsync(Project project, Cancellation
                 }
             }
 
-            private Task GetTask(Project project, Func<Task> func, CancellationToken cancellationToken)
+            [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/36158", AllowCaptures = false, Constraint = "Avoid captures to reduce GC pressure when running in the host workspace.")]
+            private Task GetTask(Project project, Func<IncrementalAnalyzer, Project, PortableExecutableReference, CancellationToken, Task> func, PortableExecutableReference reference, CancellationToken cancellationToken)
             {
                 var isRemoteWorkspace = project.Solution.Workspace.Kind == WorkspaceKind.RemoteWorkspace;
                 return isRemoteWorkspace
-                    ? Task.Run(func, cancellationToken)
-                    : func();
+                    ? GetNewTask(this, func, project, reference, cancellationToken)
+                    : func(this, project, reference, cancellationToken);
+
+                static Task GetNewTask(IncrementalAnalyzer self, Func<IncrementalAnalyzer, Project, PortableExecutableReference, CancellationToken, Task> func, Project project, PortableExecutableReference reference, CancellationToken cancellationToken)
```

Comment:
How does this help reduce allocation?

Reply:
If a method contains a lambda that captures parameters, a closure instance is created for those parameters at the start of the method whether or not the lambda is created or used. By moving the creation of the lambda to a local function, the creation of the closure instance is deferred until such time as the lambda actually needs to be created. This leaves the path above that doesn't use the lambda as a non-allocating code path.

Reply:
https://github.com/dotnet/roslyn/issues/20777